In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv("fraud test.csv")

In [3]:
data.head()

Unnamed: 0 trans_date_trans_time        cc_num  \
0           0      21/06/2020 12:14  2.291160e+15   
1           1      21/06/2020 12:14  3.573030e+15   
2           2      21/06/2020 12:14  3.598220e+15   
3           3      21/06/2020 12:15  3.591920e+15   
4           4      21/06/2020 12:15  3.526830e+15   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2     Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3  Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4    Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   

   city_pop                     job         dob  \
0    333497     Mechanical engineer  19/03/1968   
1       302  Sales professional, IT  17/01/1990   
2     34496       Librarian, public  21/10/1970   
3     54767            Set designer  25/07/1987   
4      1126      Furniture designer  06/07/1955   

                          trans_num   unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [4]:
data = data.loc[:, ~data.columns.str.contains('^Unnamed')]

In [5]:
data.head()

trans_date_trans_time        cc_num                              merchant  \
0      21/06/2020 12:14  2.291160e+15                 fraud_Kirlin and Sons   
1      21/06/2020 12:14  3.573030e+15                  fraud_Sporer-Keebler   
2      21/06/2020 12:14  3.598220e+15  fraud_Swaniawski, Nitzsche and Welch   
3      21/06/2020 12:15  3.591920e+15                     fraud_Haley Group   
4      21/06/2020 12:15  3.526830e+15                 fraud_Johnston-Casper   

         category    amt   first      last gender  \
0   personal_care   2.86    Jeff   Elliott      M   
1   personal_care  29.84  Joanne  Williams      F   
2  health_fitness  41.28  Ashley     Lopez      F   
3        misc_pos  60.05   Brian  Williams      M   
4          travel   3.19  Nathan    Massey      M   

                        street        city  ...      lat      long  city_pop  \
0            351 Darlene Green    Columbia  ...  33.9659  -80.9355    333497   
1             3638 Marsh Union     Altonah  ...  40.3207 -110.4360       302   
2         9333 Valentine Point    Bellmore  ...  40.6729  -73.5365     34496   
3  32941 Krystal Mill Apt. 552  Titusville  ...  28.5697  -80.8191     54767   
4     5783 Evan Roads Apt. 465    Falmouth  ...  44.2529  -85.0170      1126   

                      job         dob                         trans_num  \
0     Mechanical engineer  19/03/1968  2da90c7d74bd46a0caf3777415b3ebd3   
1  Sales professional, IT  17/01/1990  324cc204407e99f51b0d6ca0055005e7   
2       Librarian, public  21/10/1970  c81755dbbbea9d5c77f094348a7579be   
3            Set designer  25/07/1987  2159175b9efe66dc301f149d3d5abf8c   
4      Furniture designer  06/07/1955  57ff021bd3f328f8738bb535c302a31b   

    unix_time  merch_lat  merch_long  is_fraud  
0  1371816865  33.986391  -81.200714         0  
1  1371816873  39.450498 -109.960431         0  
2  1371816893  40.495810  -74.196111         0  
3  1371816915  28.812398  -80.883061         0  
4  1371816917  44.959148  -85.884734         0  

[5 rows x 22 columns]

In [6]:
data.shape

(555719, 22)

In [7]:
data.isna().sum()

trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 0
job                      0
dob                      0
trans_num                0
unix_time                0
merch_lat                0
merch_long               0
is_fraud                 0
dtype: int64

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 555719 entries, 0 to 555718
Data columns (total 22 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   trans_date_trans_time  555719 non-null  object 
 1   cc_num                 555719 non-null  float64
 2   merchant               555719 non-null  object 
 3   category               555719 non-null  object 
 4   amt                    555719 non-null  float64
 5   first                  555719 non-null  object 
 6   last                   555719 non-null  object 
 7   gender                 555719 non-null  object 
 8   street                 555719 non-null  object 
 9   city                   555719 non-null  object 
 10  state                  555719 non-null  object 
 11  zip                    555719 non-null  int64  
 12  lat                    555719 non-null  float64
 13  long                   555719 non-null  float64
 14  city_pop               555719 non-nu

In [9]:
data['is_fraud'].value_counts() #very unbalanced dataset so need to balance it

0    553574
1      2145
Name: is_fraud, dtype: int64

In [10]:
data.groupby('is_fraud').mean()

cc_num         amt           zip        lat       long  \
is_fraud                                                                 
0         4.179643e+17   67.614408  48846.424285  38.541406 -90.230495   
1         3.854275e+17  528.356494  47862.900233  39.019971 -90.445556   

              city_pop     unix_time  merch_lat  merch_long  
is_fraud                                                     
0         88313.692365  1.380683e+09  38.540954  -90.230474  
1         64529.321678  1.379587e+09  39.018618  -90.465222

In [11]:
legit = data[data.is_fraud == 0]
fraud = data[data.is_fraud == 1]

In [12]:
legit_sample = legit.sample(n=2145)

In [13]:
merged_card_data = pd.concat([legit_sample, fraud], axis = 0)

In [14]:
merged_card_data.shape

(4290, 22)

In [15]:
merged_card_data['is_fraud'].value_counts()

0    2145
1    2145
Name: is_fraud, dtype: int64

In [16]:
merged_card_data['gender'].value_counts()

F    2329
M    1961
Name: gender, dtype: int64

In [17]:
merged_card_data.groupby('is_fraud').mean()

cc_num         amt           zip        lat       long  \
is_fraud                                                                 
0         4.470426e+17   66.310196  48897.248485  38.643115 -90.128349   
1         3.854275e+17  528.356494  47862.900233  39.019971 -90.445556   

              city_pop     unix_time  merch_lat  merch_long  
is_fraud                                                     
0         86978.216783  1.380715e+09  38.658757  -90.142401  
1         64529.321678  1.379587e+09  39.018618  -90.465222

In [18]:
merged_card_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4290 entries, 408756 to 517571
Data columns (total 22 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   trans_date_trans_time  4290 non-null   object 
 1   cc_num                 4290 non-null   float64
 2   merchant               4290 non-null   object 
 3   category               4290 non-null   object 
 4   amt                    4290 non-null   float64
 5   first                  4290 non-null   object 
 6   last                   4290 non-null   object 
 7   gender                 4290 non-null   object 
 8   street                 4290 non-null   object 
 9   city                   4290 non-null   object 
 10  state                  4290 non-null   object 
 11  zip                    4290 non-null   int64  
 12  lat                    4290 non-null   float64
 13  long                   4290 non-null   float64
 14  city_pop               4290 non-null   int64  
 1

###### Dropping unnecessary columns

In [19]:
merged_card_data.drop(columns = ['trans_date_trans_time', 'first', 'last', 'street', 'city_pop', 'unix_time', 'trans_num'], axis = 1, inplace = True)

In [20]:
merged_card_data.shape

(4290, 15)

###### Feature Engineering

In [21]:
merged_card_data['location'] = merged_card_data['city'] + ', ' + merged_card_data['state'] + ', ' + merged_card_data['zip'].astype(str)
merged_card_data = merged_card_data.drop(['city', 'state', 'zip'], axis=1)

In [22]:
merged_card_data['lat_long'] = merged_card_data['lat'].astype(str) + ', ' + merged_card_data['long'].astype(str)
merged_card_data = merged_card_data.drop(['lat', 'long'], axis=1)
merged_card_data.head()

cc_num                            merchant      category  \
408756  3.586010e+15      fraud_O'Reilly, Mohr and Purdy          home   
141814  2.224100e+15                      fraud_Cole PLC   grocery_pos   
454418  6.304520e+11                   fraud_Mante Group      misc_net   
98771   3.567880e+15                  fraud_Hickle Group  shopping_pos   
31578   4.587580e+15  fraud_Reichert, Shanahan and Hayes  shopping_net   

           amt gender               job         dob  merch_lat  merch_long  \
408756   17.20      F         Osteopath  29/02/1984  47.266159 -122.886556   
141814  167.36      F   Naval architect  11/09/1963  31.223798  -81.559031   
454418   33.40      F      Immunologist  12/06/1972  47.118586  -86.219648   
98771    30.28      F   Producer, radio  14/08/1958  44.702999  -75.915299   
31578    43.11      F  Financial trader  12/04/1976  43.566546  -70.051878   

        is_fraud               location            lat_long  
408756         0    Issaquah, WA, 98027  47.4974, -122.0107  
141814         0   Kings Bay, GA, 31547   30.7906, -81.5607  
454418         0     Wetmore, MI, 49895   46.3535, -86.6345  
98771          0  Brownville, NY, 13615   44.0577, -76.0196  
31578          0    Nobleboro, ME, 4555   44.0943, -69.4828

In [23]:
merged_card_data['merchant_lat_long'] = merged_card_data['merch_lat'].astype(str) + ', ' + merged_card_data['merch_long'].astype(str)
merged_card_data = merged_card_data.drop(['merch_lat', 'merch_long'], axis=1)
merged_card_data.head()

cc_num                            merchant      category  \
408756  3.586010e+15      fraud_O'Reilly, Mohr and Purdy          home   
141814  2.224100e+15                      fraud_Cole PLC   grocery_pos   
454418  6.304520e+11                   fraud_Mante Group      misc_net   
98771   3.567880e+15                  fraud_Hickle Group  shopping_pos   
31578   4.587580e+15  fraud_Reichert, Shanahan and Hayes  shopping_net   

           amt gender               job         dob  is_fraud  \
408756   17.20      F         Osteopath  29/02/1984         0   
141814  167.36      F   Naval architect  11/09/1963         0   
454418   33.40      F      Immunologist  12/06/1972         0   
98771    30.28      F   Producer, radio  14/08/1958         0   
31578    43.11      F  Financial trader  12/04/1976         0   

                     location            lat_long       merchant_lat_long  
408756    Issaquah, WA, 98027  47.4974, -122.0107  47.266159, -122.886556  
141814   Kings Bay, GA, 31547   30.7906, -81.5607   31.223798, -81.559031  
454418     Wetmore, MI, 49895   46.3535, -86.6345   47.118586, -86.219648  
98771   Brownville, NY, 13615   44.0577, -76.0196   44.702999, -75.915299  
31578     Nobleboro, ME, 4555   44.0943, -69.4828   43.566546, -70.051878

###### Label Encoder

In [24]:
from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
label_encoder = LabelEncoder()

# Fit label encoder and return encoded labels
merged_card_data['gender'] = label_encoder.fit_transform(merged_card_data['gender'])
merged_card_data['job'] = label_encoder.fit_transform(merged_card_data['job'])
merged_card_data['dob'] = label_encoder.fit_transform(merged_card_data['dob'])
merged_card_data['cc_num'] = label_encoder.fit_transform(merged_card_data['cc_num'])
merged_card_data['merchant'] = label_encoder.fit_transform(merged_card_data['merchant'])
merged_card_data['category'] = label_encoder.fit_transform(merged_card_data['category'])
merged_card_data['location'] = label_encoder.fit_transform(merged_card_data['location'])
merged_card_data['lat_long'] = label_encoder.fit_transform(merged_card_data['lat_long'])
merged_card_data['merchant_lat_long'] = label_encoder.fit_transform(merged_card_data['merchant_lat_long'])



In [25]:
merged_card_data.head()

cc_num  merchant  category     amt  gender  job  dob  is_fraud  \
408756     498       457         6   17.20       0  295  735         0   
141814     325        94         4  167.36       0  280  267         0   
454418      39       402         8   33.40       0  222  289         0   
98771      475       252        12   30.28       0  322  364         0   
31578      594       500        11   43.11       0  178  285         0   

        location  lat_long  merchant_lat_long  
408756       361       788               4125  
141814       381        57                313  
454418       786       777               4115  
98771        105       708               3804  
31578        522       709               3588

###### Seperating features and labels

In [26]:
X = merged_card_data.drop(columns = 'is_fraud', axis = 1)
Y = merged_card_data['is_fraud']
print(X,Y)

        cc_num  merchant  category      amt  gender  job  dob  location  \
408756     498       457         6    17.20       0  295  735       361   
141814     325        94         4   167.36       0  280  267       381   
454418      39       402         8    33.40       0  222  289       786   
98771      475       252        12    30.28       0  322  364       105   
31578      594       500        11    43.11       0  178  285       522   
...        ...       ...       ...      ...     ...  ...  ...       ...   
517197     332       661        12  1041.51       1  103   65       669   
517274     332       347        12   868.09       1  103   65       669   
517341     332       439        11  1039.42       1  103   65       669   
517529     332        29         4   289.27       1  103   65       669   
517571     332       281         8   766.38       1  103   65       669   

        lat_long  merchant_lat_long  
408756       788               4125  
141814        57       

In [27]:
X.shape

(4290, 10)

In [28]:
Y.shape

(4290,)

###### Train Test Split

In [29]:
from sklearn.model_selection import train_test_split

In [30]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = .2, stratify = Y)

###### StandardScaler

In [31]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
print(X.shape, X_train_scaled.shape, X_test_scaled.shape)

(4290, 10) (3432, 10) (858, 10)


###### Building NN

In [32]:
#importing Tensorflow and Keras
import tensorflow as tf
from tensorflow import keras

In [33]:
#setting up layers of neural network
model = keras.Sequential([
    keras.layers.Flatten(input_shape = (10,)), 
    keras.layers.Dense(100, activation = 'relu', kernel_regularizer=tf.keras.regularizers.l2(0.01)),
    tf.keras.layers.Dropout(0.5),
    keras.layers.Dense(20, activation = 'relu', kernel_regularizer=tf.keras.regularizers.l2(0.01)),
    tf.keras.layers.Dropout(0.5),
    keras.layers.Dense(2, activation = 'sigmoid', kernel_regularizer=tf.keras.regularizers.l2(0.01))

])
# Print model summary
model.summary()

C:\Users\Zooby\anaconda3\lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 10)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 100)            │         1,100 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 100)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 20)             │         2,020 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 20)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 2)              │            42 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,162 (12.35 KB)

 Trainable params: 3,162 (12.35 KB)

 Non-trainable params: 0 (0.00 B)

In [34]:
model.compile(optimizer = 'adam',
              loss = 'sparse_categorical_crossentropy',
              metrics = ['accuracy']
             )

In [35]:
history = model.fit(X_train_scaled, Y_train, validation_split= .1, epochs = 10) 

Epoch 1/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5551 - loss: 1.2281 - val_accuracy: 0.7297 - val_loss: 0.9446
Epoch 2/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 0s 868us/step - accuracy: 0.7048 - loss: 0.9192 - val_accuracy: 0.7791 - val_loss: 0.7969
Epoch 3/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 0s 856us/step - accuracy: 0.7827 - loss: 0.7532 - val_accuracy: 0.8052 - val_loss: 0.7098
Epoch 4/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 0s 857us/step - accuracy: 0.8175 - loss: 0.6611 - val_accuracy: 0.7994 - val_loss: 0.6534
Epoch 5/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 0s 925us/step - accuracy: 0.8322 - loss: 0.6147 - val_accuracy: 0.8023 - val_loss: 0.6141
Epoch 6/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 0s 904us/step - accuracy: 0.8484 - loss: 0.5604 - val_accuracy: 0.7994 - val_loss: 0.5846
Epoch 7/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 0s 914us/step - accuracy: 0.8365 - loss: 0.5457 - val_accuracy: 0.7994 - val_loss: 0.5693
Epoch 8/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 0s 904us/step - accuracy: 0.8457 - loss: 0.5109 - val_accuracy: 0.80

In [36]:
score, acc = model.evaluate(X_test_scaled, Y_test)
print('Test Loss =', score)
print('Test Accuracy =', acc)

27/27 ━━━━━━━━━━━━━━━━━━━━ 0s 652us/step - accuracy: 0.8655 - loss: 0.4777
Test Loss = 0.4739050269126892
Test Accuracy = 0.8554778695106506


###### Building other ML models for better performance

In [37]:
from sklearn.linear_model import LogisticRegression
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, confusion_matrix 
from sklearn import svm
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

# Define parameter grids for randomized search (coarse search)
logistic_param_grid_coarse = {
    'penalty': ['l1', 'l2'],
    'C': [0.01, 0.1, 1, 10, 100],
    'solver': ['liblinear', 'saga']
}

xgboost_param_grid_coarse = {
    'learning_rate': [0.01, 0.1, 0.2],
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 7],
    'min_child_weight': [1, 3, 5],
    'subsample': [0.5, 0.8, 1.0],
    'colsample_bytree': [0.5, 0.8, 1.0]
}

svm_param_grid_coarse = {
    'C': [0.1, 1, 10],
    'gamma': [0.01, 0.1, 1],
    'kernel': ['linear', 'rbf']
}

# RandomizedSearchCV for each model (coarse search)
logistic_random_search_coarse = RandomizedSearchCV(LogisticRegression(), logistic_param_grid_coarse, n_iter=30, cv=5, n_jobs=-1)
xgboost_random_search_coarse = RandomizedSearchCV(XGBClassifier(objective='binary:logistic'), xgboost_param_grid_coarse, n_iter=30, cv=5, n_jobs=-1)
svm_random_search_coarse = RandomizedSearchCV(svm.SVC(), svm_param_grid_coarse, n_iter=30, cv=5, n_jobs=-1)

# Fit models using RandomizedSearchCV (coarse search)
logistic_random_search_coarse.fit(X_train_scaled, Y_train)
xgboost_random_search_coarse.fit(X_train_scaled, Y_train)
svm_random_search_coarse.fit(X_train_scaled, Y_train)

# Get best hyperparameters from RandomizedSearchCV (coarse search)
best_logistic_params_coarse = logistic_random_search_coarse.best_params_
best_xgboost_params_coarse = xgboost_random_search_coarse.best_params_
best_svm_params_coarse = svm_random_search_coarse.best_params_

# Define parameter grids for GridSearchCV (fine search)
logistic_param_grid_fine = {
    'penalty': [best_logistic_params_coarse['penalty']],
    'C': [best_logistic_params_coarse['C'] * i for i in [0.1, 1, 10]],
    'solver': [best_logistic_params_coarse['solver']]
}

xgboost_param_grid_fine = {
    'learning_rate': [best_xgboost_params_coarse['learning_rate'] * i for i in [0.5, 1, 2]],
    'n_estimators': [best_xgboost_params_coarse['n_estimators']],
    'max_depth': [best_xgboost_params_coarse['max_depth']],
    'min_child_weight': [best_xgboost_params_coarse['min_child_weight']],
    'subsample': [best_xgboost_params_coarse['subsample']],
    'colsample_bytree': [best_xgboost_params_coarse['colsample_bytree']]
}

svm_param_grid_fine = {
    'C': [best_svm_params_coarse['C'] * i for i in [0.1, 1, 10]],
    'gamma': [best_svm_params_coarse['gamma'] * i for i in [0.1, 1, 10]],
    'kernel': [best_svm_params_coarse['kernel']]
}

# GridSearchCV for each model (fine search)
logistic_grid_search_fine = GridSearchCV(LogisticRegression(), param_grid=logistic_param_grid_fine, cv=5, n_jobs=-1)
xgboost_grid_search_fine = GridSearchCV(XGBClassifier(objective='binary:logistic'), param_grid=xgboost_param_grid_fine, cv=5, n_jobs=-1)
svm_grid_search_fine = GridSearchCV(svm.SVC(), param_grid=svm_param_grid_fine, cv=5, n_jobs=-1)

# Fit models using GridSearchCV (fine search)
logistic_grid_search_fine.fit(X_train_scaled, Y_train)
xgboost_grid_search_fine.fit(X_train_scaled, Y_train)
svm_grid_search_fine.fit(X_train_scaled, Y_train)

# Print best hyperparameters from GridSearchCV (fine search)
print("Logistic Regression Best Parameters (Fine Search):", logistic_grid_search_fine.best_params_)
print("XGBoost Best Parameters (Fine Search):", xgboost_grid_search_fine.best_params_)
print("SVM Best Parameters (Fine Search):", svm_grid_search_fine.best_params_)

# Compare cross-validated scores of each model
logistic_cv_score_fine = logistic_grid_search_fine.best_score_
xgboost_cv_score_fine = xgboost_grid_search_fine.best_score_
svm_cv_score_fine = svm_grid_search_fine.best_score_

# Select the best model based on cross-validated scores
best_model_fine = None
if logistic_cv_score_fine >= xgboost_cv_score_fine and logistic_cv_score_fine >= svm_cv_score_fine:
    best_model_fine = logistic_grid_search_fine.best_estimator_
elif xgboost_cv_score_fine >= logistic_cv_score_fine and xgboost_cv_score_fine >= svm_cv_score_fine:
    best_model_fine = xgboost_grid_search_fine.best_estimator_
else:
    best_model_fine = svm_grid_search_fine.best_estimator_

# Evaluate the best model on the test set
train_accuracy_fine = best_model_fine.score(X_train_scaled, Y_train)
print("Best Model Train Accuracy (Fine Search):", train_accuracy_fine)
test_accuracy_fine = best_model_fine.score(X_test_scaled, Y_test)
print("Best Model Test Accuracy (Fine Search):", test_accuracy_fine)




C:\Users\Zooby\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 20 is smaller than n_iter=30. Running 20 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(
C:\Users\Zooby\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 18 is smaller than n_iter=30. Running 18 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Logistic Regression Best Parameters (Fine Search): {'C': 0.01, 'penalty': 'l1', 'solver': 'saga'}
XGBoost Best Parameters (Fine Search): {'colsample_bytree': 0.5, 'learning_rate': 0.2, 'max_depth': 5, 'min_child_weight': 3, 'n_estimators': 100, 'subsample': 1.0}
SVM Best Parameters (Fine Search): {'C': 1, 'gamma': 0.1, 'kernel': 'rbf'}
Best Model Train Accuracy (Fine Search): 0.9947552447552448
Best Model Test Accuracy (Fine Search): 0.9627039627039627


In [38]:
y_pred = best_model_fine.predict(X_test_scaled)

# Calculate evaluation metrics
accuracy = accuracy_score(Y_test, y_pred)
precision = precision_score(Y_test, y_pred)
recall = recall_score(Y_test, y_pred)
f1 = f1_score(Y_test, y_pred)
roc_auc = roc_auc_score(Y_test, best_model_fine.predict_proba(X_test_scaled)[:, 1])
conf_matrix = confusion_matrix(Y_test, y_pred)

# Print the results
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)
print("ROC AUC Score:", roc_auc)
print("Confusion Matrix:")
print(conf_matrix)

Accuracy: 0.9627039627039627
Precision: 0.9627039627039627
Recall: 0.9627039627039627
F1 Score: 0.9627039627039627
ROC AUC Score: 0.9929798251476574
Confusion Matrix:
[[413  16]
 [ 16 413]]


# Final Notes:

Neural Network model accuracy on validation set: 86%

ML best model using GridSearchCV accuracy on test set: 96%